From: "Liam R. Howlett" <Liam.Howlett@oracle.com>
Subject: maple_tree: Fix potential insufficient nodes on mas_spanning_rebalance()

When a spanning store occurs at a non-root node but causes an insufficient
leave, mas_spanning_rebalance() was not detecting it as a non-root node. 
This only happened when the spanning write was detected at the root node.

Link: https://lkml.kernel.org/r/20220420134737.yuhkl6znauwpp7j5@revolver
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/maple_tree.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/lib/maple_tree.c~maple-tree-add-new-data-structure-fix-2
+++ a/lib/maple_tree.c
@@ -2894,7 +2894,7 @@ static int mas_spanning_rebalance(struct
 	mast->free = &free;
 	mast->destroy = &destroy;
 	l_mas.node = r_mas.node = m_mas.node = MAS_NONE;
-	if (!mas_is_root_limits(mas) &&
+	if (!mas_is_root_limits(mast->orig_l) &&
 	    unlikely(mast->bn->b_end <= mt_min_slots[mast->bn->type])) {
 		/*
 		 * Do not free the current node as it may be freed in a bulk
_
